### PR TITLE
[aes] Remove hierarchical paths in DOM S-Box

### PR DESCRIPTION
**Pull request by vogelpi**
This resolves lowRISC/OpenTitan#10180.

### CODE REVIEW

**Comment by vogelpi** [comment]
Please don't merge this yet, I am currently running formal verification - just to be on the safe side.

**Comment by vogelpi** [comment]
Update, after some more thinking and analyzing the number of flops ending up in the list I realized that:
1. By moving the instantiation of the pipeline registers to the outside of the DOM-dep multipliers, we can enable the sharing of pipeline registers between the DOM-dep multipliers in Stage 3. This saves another 4 FFs per DOM S-Box and the design becomes slightly more readable.
2. We should also use `prim_flops` for the PRD buffering. Otherwise synthesis tools might optimize away some of these flops because they realize that the downstream circuit only evaluates their outputs in one particular clock cycle. This will introduce unnecessary SCA leakage. (Vivado luckily doesn't do this).
3. The keep_hierarchy constraints in the Yosys synthesis script should be applied to all primitives, not just prim_bufs.

@msfschaffner do you have an easy way of identifying how many FFs DC ends up with? We should check this number. With the last version it's 147 FFs per S-Box.

File: hw/ip/aes/rtl/aes_sbox_dom.sv
Comment (msfschaffner) [inline hw/ip/aes/rtl/aes_sbox_dom.sv:946]: nit: you can also use `_q1` `_q2` etc so that these do not get endlessly long ;)

File: hw/ip/aes/pre_syn/tcl/yosys_run_synth.tcl
Comment (msfschaffner) [inline hw/ip/aes/pre_syn/tcl/yosys_run_synth.tcl:36]: if you want to more accurately model this, you could place this constraint only on the prims that we actually constrain:
https://github.com/lowRISC/opentitan/blob/72643363e818f11a292d87913b7b3acc0e002714/hw/syn/tools/dc/gtech-constraints.sdc#L12-L18

otherwise it will affect other prims as well (granted, it does not apply to all prims since it is technology specific, but still).

**Review by msfschaffner — APPROVED** [review]
Thanks for the cleanup @vogelpi.

I am merging this to unblock things.

If there is any additional feedback on this PR, we can handle that in follow-ups.

File: hw/ip/aes/pre_syn/tcl/yosys_run_synth.tcl
Comment (vogelpi) [reply hw/ip/aes/pre_syn/tcl/yosys_run_synth.tcl:36]: Oh, that's a good point. Right now, and for AES it's okay because in AES we just use `xor`, `buf`, `flop`, and `flop_en` but no other Xilinx/generic prims like memories etc. But I should fix this at some point.

**Pull request merged**
